### ISSUE TITLE
sequencer, mempool: make transaction execution and CheckTx share code

### ISSUE THREAD

**Issue by Lilyjjo** [astriaorg/astria#1775]
We want the mempool's CheckTx and the app's transaction execution logic to mostly perform the same checks. Right now the mempool duplicates a lot of the checks that execution performs, which can lead to issues if devs forget to update the CheckTx code when changing the execution code. Instead of trying to remember to update code in a certain way, we should have the code be shared between the two locations.

For example, CheckTx relies on the function [`get_total_transaction_cost()`](https://github.com/astriaorg/astria/blob/5fc9c56b6030038e5d7806a6785adee8a2d18cd3/crates/astria-sequencer/src/transaction/checks.rs#L68), which has a manual rewrite of fee logic that is trying to duplicate code in `check_and_execute()` which does the [action's non-fee balance reductions](https://github.com/astriaorg/astria/blob/5fc9c56b6030038e5d7806a6785adee8a2d18cd3/crates/astria-sequencer/src/accounts/action.rs#L63) inside of the action execution. This getting out of sync could cause issues in the mempool's pending vs parked distinction. (Note: the `check_and_execute()` code also calls into the manual function which is a redundant check that should be removed). 

This logic was initially separated due to the performance concerns of running the full transaction execution in CheckTx. Since then, we've changed the mempool to only run CheckTx on the first time a transaction is seen instead of on every block. For performance, it would be nice if we could have the CheckTx implementation run the actions checks in parallel. 

Other implementation notes:
- CheckTx has slightly different requirements than actual transaction execution. CheckTx doesn't want to reject transactions that have too-high nonces or lack balance, and, CheckTx wants to know the total cost of the transaction. The shared implementation should be written in a way where all shared failure checks run first and be able to return the total cost of the transaction even when the account has a too low nonce or not enough balance.

┆Issue Number: ENG-976

**Comment by joroshiba** [comment]
This issue is stale because it has been open 45 days with no activity. Remove stale label or this issue
be closed in 7 days.

**Comment by joroshiba** [comment]
This issue was closed because it was stale
